### PR TITLE
Add context lost event and context restore event for canvas 2d rendering context

### PR DESCRIPTION
**Pull request by yiyix**
Placeholder for context lost/restore event.

Description:
https://github.com/fserb/canvas2D/blob/master/spec/context-loss.md

Working on whatwg spec now.

### CODE REVIEW

File: content/en-US/api/2dcontextlostevent.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+**When this feature ships, the content below will live on MDN under
+[developer.mozilla.org/en-US/docs/Web/API](https://developer.mozilla.org/en-US/docs/Web/API).**
+
+## Description
+
+The 2d Rendering context losses may be intended by the user agent (to resolve
+resource contention), or may be forced by external factors (e.g. a graphics
+driver reset). In both scenarios, this event will be triggered after the 
+2dRenderingContext has been lost.
```
Comment (mysteryDate) [inline content/en-US/api/2dcontextlostevent.md]: This reads a little strangely. I think you should focus first on what CanvasRendering2DContextLost _is_ and what it _does_. Something along the lines of:
  "`contextLost` is a javascript event that is triggered when a 2d context on the page has been "lost". Contexts can be lost when a driver crashes, the application runs out of memory..." etc, etc. 
    Then maybe describe why they're useful?

File: content/en-US/api/2dcontextlostevent.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+const canvas = document.getElementById('canvas');
+const ctx = canvas.getContext('2d');
+
+canvas.addEventListener('contextlost', (event) => {
+  console.log(event);
+});
+
+// Chrome allows users to force canvas context to be lost.
+window.internals.forceLoseCanvasContext(canvas, "2d");
```
Comment (mysteryDate) [inline content/en-US/api/2dcontextlostevent.md]: Isn't this only allowable with an extension?

File: content/en-US/api/2dcontextlostevent.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+// After this rendering context 'ctx' is lost, "contextlost" is logged.
```
Comment (mysteryDate) [inline content/en-US/api/2dcontextlostevent.md]: The text "contextlost" is not logged, an event is.

File: content/en-US/api/2dcontextrestoreevent.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+---
+
+**When this feature ships, the content below will live on MDN under
+[developer.mozilla.org/en-US/docs/Web/API](https://developer.mozilla.org/en-US/docs/Web/API).**
+
+## Description
+
+The 2d Rendering context losses may be intended by the user agent (to resolve
+resource contention), or may be forced by external factors (e.g. a graphics
+driver reset). If the context is lost then restored, this event will be triggered.
```
Comment (mysteryDate) [inline content/en-US/api/2dcontextrestoreevent.md]: Same as above, I feel like you need to start this explanation from somewhere more basic.

File: content/en-US/api/2dcontextrestoreevent.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+## Description
+
+The 2d Rendering context losses may be intended by the user agent (to resolve
+resource contention), or may be forced by external factors (e.g. a graphics
+driver reset). If the context is lost then restored, this event will be triggered.  
+
+Once the context is restored, the resources such as drawings that were created before
+the context was lost are no longer valid. You need to reinitialize the state of your 
+context and recreate resources.
```
Comment (mysteryDate) [inline content/en-US/api/2dcontextrestoreevent.md:20]: This reads well and clear though.

File: content/en-US/api/2dcontextrestoreevent.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+canvas.addEventListener('contextrestored', (event) => {
+  console.log(event);
+});
+
+// Chrome allows users to force canvas context to be lost.
+window.internals.forceLoseCanvasContext(canvas, "2d");
+
+// After this rendering context 'ctx' is lost, 'ctx' will be restored automatically.
+// "contextrestored" is logged.
```
Comment (mysteryDate) [inline content/en-US/api/2dcontextrestoreevent.md]: Same comments as above for this code.

File: content/en-US/api/2dcontextlostevent.md (hunk 1 above)
Comment (jpmedley) [reply content/en-US/api/2dcontextlostevent.md]: He's right except that MDN has a standard way to do this for interfaces:

>The `2dcontextlost` interface of the _someAPI_ _someVerb_ ...

File: content/en-US/api/2dcontextlostevent.md (hunk 2 above)
Comment (jpmedley) [reply content/en-US/api/2dcontextlostevent.md]: I was wondering the same thing.

Is this event web platform or something related to extensions?

File: content/en-US/api/2dcontextrestoreevent.md (hunk 4 above)
Comment (jpmedley) [reply content/en-US/api/2dcontextrestoreevent.md]: +1

File: content/en-US/api/2dcontextrestoreevent.md (hunk 6 above)
Comment (jpmedley) [reply content/en-US/api/2dcontextrestoreevent.md]: +1

File: content/en-US/api/2dcontextlostevent.md (hunk 1 above)
Comment (yiyix) [reply content/en-US/api/2dcontextlostevent.md]: I wonder if i need to add why they are useful. I mean we don't have it for other canvas rendering 2d methods or webgl contest lose event. I feel users will find their way to use this function. If you still believe it's better to have it, I will add it later.

File: content/en-US/api/2dcontextlostevent.md (hunk 2 above)
Comment (yiyix) [reply content/en-US/api/2dcontextlostevent.md]: it's visible with flag "--expose-internals-for-testing". I think it's best to remove this line.... Maybe it's not a good idea to share these functions with public.

File: content/en-US/api/2dcontextlostevent.md (hunk 3 above)
Comment (yiyix) [reply content/en-US/api/2dcontextlostevent.md]: updated.

File: content/en-US/api/2dcontextrestoreevent.md (hunk 4 above)
Comment (yiyix) [reply content/en-US/api/2dcontextrestoreevent.md]: I change it so it's the same style as the context lost event.

File: content/en-US/api/2dcontextrestoreevent.md (hunk 6 above)
Comment (yiyix) [reply content/en-US/api/2dcontextrestoreevent.md]: updated.

File: content/en-US/api/2dcontextlostevent.md [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+---
+
+**When this feature ships, the content below will live on MDN under
+[developer.mozilla.org/en-US/docs/Web/API](https://developer.mozilla.org/en-US/docs/Web/API).**
+
+## Description
+
+The `contextRestored` event is a javascript event that is triggered when the user agent detects
+that the context associated with 'CanvasRenderingContext2D' on the page is "lost". Contexts can
+be lost for several reasons, such as a driver crashes, the application runs out of memory, etc.
```
Comment (jpmedley) [inline content/en-US/api/2dcontextlostevent.md]: At the risk of sounding like a pest, this has to follow the MDN wording *exactly*. The API section is a reference like a dictionary. Structural and lexical consistency support quick scanning for information. [Here's an example of what the wording should like.](https://developer.mozilla.org/en-US/docs/Web/API/MIDIPort/onstatechange)

File: content/en-US/api/2dcontextrestoreevent.md [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+---
+
+**When this feature ships, the content below will live on MDN under
+[developer.mozilla.org/en-US/docs/Web/API](https://developer.mozilla.org/en-US/docs/Web/API).**
+
+## Description
+
+The `contextRestored` event is a javascript event that is triggered when the user agent
+detects that the context associated with 'CanvasRenderingContext2D' on the page is "restored"
+after being "lost".
```
Comment (jpmedley) [inline content/en-US/api/2dcontextrestoreevent.md]: Here's [how an interface should be worded](https://developer.mozilla.org/en-US/docs/Web/API/MIDIPort).

**Review by jpmedley — CHANGES_REQUESTED** [review]

File: content/en-US/api/2dcontextlostevent.md [hunk 9]
```diff
@@ -11,9 +11,11 @@ browser_compatibility: api.2dcontextlost
 
 ## Description
 
-The `contextRestored` event is a javascript event that is triggered when the user agent detects
-that the context associated with 'CanvasRenderingContext2D' on the page is "lost". Contexts can
-be lost for several reasons, such as a driver crashes, the application runs out of memory, etc.
+The `contextlost` `EventHandler` of [Canvas Api](https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API)
+is an [`EventHandler`](https://developer.mozilla.org/en-US/docs/Web/Events/Event_handlers) that
+is triggered when the user agent detects
```
Comment (jpmedley) [inline content/en-US/api/2dcontextlostevent.md]:
```suggestion
The `contextlost` [`EventHandler`](https://developer.mozilla.org/en-US/docs/Web/Events/Event_handlers) of [Canvas Api](https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API)
is triggered when the user agent detects
```

File: content/en-US/api/2dcontextrestoreevent.md [hunk 10]
```diff
@@ -11,9 +11,10 @@ browser_compatibility: api.2dcontextlost
 
 ## Description
 
-The `contextRestored` event is a javascript event that is triggered when the user agent
-detects that the context associated with 'CanvasRenderingContext2D' on the page is "restored"
-after being "lost".
+The `contextrestored` `EventHandler` of [Canvas Api](https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API)
+is an [`EventHandler`](https://developer.mozilla.org/en-US/docs/Web/Events/Event_handlers) that
+is triggered when the user agent "restores" the backing storage associated with
```
Comment (jpmedley) [inline content/en-US/api/2dcontextrestoreevent.md]:
```suggestion
The `contextrestored` [`EventHandler`](https://developer.mozilla.org/en-US/docs/Web/Events/Event_handlers) of [Canvas Api](https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API)
is triggered when the user agent "restores" the backing storage associated with
```

**Review by jpmedley — CHANGES_REQUESTED** [review]

**Comment by yiyix** [comment]
Thank you for the review!

**Review by jpmedley — APPROVED** [review]

**Pull request merged**